core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -1,187 +1,337 @@
 use {
-    itertools::Itertools,
+    crate::replay_stage::DUPLICATE_THRESHOLD,
     solana_gossip::{
         cluster_info::ClusterInfo, contact_info::ContactInfo, crds::Cursor, epoch_slots::EpochSlots,
     },
-    solana_runtime::{bank::Bank, epoch_stakes::NodeIdToVoteAccounts},
-    solana_sdk::{
-        clock::{Slot, DEFAULT_SLOTS_PER_EPOCH},
-        pubkey::Pubkey,
-        timing::AtomicInterval,
-    },
+    solana_sdk::{clock::Slot, pubkey::Pubkey, timing::AtomicInterval},
     std::{
-        collections::{BTreeMap, HashMap},
-        sync::{Arc, Mutex, RwLock},
+        collections::{HashMap, VecDeque},
+        hash::RandomState,
+        ops::Range,
+        sync::{
+            atomic::{AtomicU64, Ordering},
+            Arc, Mutex, RwLock,
+        },
     },
 };
 
 // Limit the size of cluster-slots map in case
 // of receiving bogus epoch slots values.
-const CLUSTER_SLOTS_TRIM_SIZE: usize = 524_288; // 512K
+// This also constraints the size of the datastructure
+// if we are really really far behind.
+const CLUSTER_SLOTS_TRIM_SIZE: usize = 50000;
+
+pub type Stake = u64;
+
+//This is intended to be switched to solana_pubkey::PubkeyHasherBuilder
+type PubkeyHasherBuilder = RandomState;
+pub(crate) type ValidatorStakesMap = HashMap<Pubkey, Stake, PubkeyHasherBuilder>;
 
-pub(crate) type SlotPubkeys = HashMap</*node:*/ Pubkey, /*stake:*/ u64>;
+/// Pubkey-stake map for nodes that have confirmed this slot.
+/// If stake is zero the node has not confirmed the slot.
+pub(crate) type SlotPubkeys =
+    HashMap<Pubkey, AtomicU64 /* stake supporting */, PubkeyHasherBuilder>;
+
+/// Amount of stake that needs to lock the slot for us to stop updating it.
+/// This must be above `DUPLICATE_THRESHOLD` but also high enough to not starve
+/// repair (as it will prefer nodes that have confirmed a slot)
+const FREEZE_THRESHOLD: f64 = 0.9;
+// whatever freeze threshold we should be above DUPLICATE_THRESHOLD in order to not break consensus.
+static_assertions::const_assert!(FREEZE_THRESHOLD > DUPLICATE_THRESHOLD * 1.1);
 
 #[derive(Default)]
 pub struct ClusterSlots {
-    cluster_slots: RwLock<BTreeMap<Slot, Arc<RwLock<SlotPubkeys>>>>,
-    validator_stakes: RwLock<Arc<NodeIdToVoteAccounts>>,
-    epoch: RwLock<Option<u64>>,
-    cursor: Mutex<Cursor>,
-    last_report: AtomicInterval,
+    // ring buffer storing, per slot, which stakes were committed to a certain slot.
+    cluster_slots: RwLock<VecDeque<RowContent>>,
+    // a cache of validator stakes for reuse internally, updated at epoch boundary.
```

Comment:
Switching to a VecDeque from BTreeMap allows us to have O(1) lookup and easy reuse of the rows in the datastructure to avoid malloc pressure